### DISCUSSION TITLE
Prioritize TypeScript-compatible imports for `.js` and `.ts` files

### DISCUSSION THREAD

**Discussion by Princesseuh** [withastro/roadmap#191] [Proposal]
# Background

Currently in an Astro file, the following ways of importing from TS files inside the frontmatter are supported:

```astro
---
import { thing } from "../scripts/tsscript" // TypeScript recommended way of doing things
import { thing } from "../scripts/tsscript.ts" // Non-standard way of importing TS files
---
```

We'll focus on the later here. In the case of Astro, this is not a problem because we don't really mind the extension being there. This is bothersome for one thing: **Editor support** (my thing!). Our editor support is directly based on TypeScript and TypeScript heavily discourage using a `.ts` import.

When using an import with a `.ts` file extension, TypeScript language features all stops working for that import and TypeScript show an error telling the user to not use an extension. We currently suppress that error inside our language tooling since Astro does support `.ts`, which leads to [users sending feedback](https://github.com/withastro/language-tools/issues/126) that some features are not working when they actually do work without the file extension

Additionally, inside script tags Astro currently absolutely requires the file extension (in build only, dev works though? Not sure what our stance is), which mean that it's impossible to benefits from TypeScript features for imports there while still having a working product in Astro

```astro
<!-- This crash Astro in build, the file extension is required inside script tags -->
<script>
  import { thing } from "../scripts/tsscript"
</script>
```

## TL;DR

`.ts` file extension, no hover info, no errors, no TypeScript features, nothing, rien, nada

![image](https://user-images.githubusercontent.com/3019731/166801734-f4dad823-4ee1-412d-8803-02517373f7e5.png)

No file extension, cool hover info, error for missing argument, TypeScript features

![image](https://user-images.githubusercontent.com/3019731/166801577-c1dd72c6-aac5-4b7b-b3d5-5475e27bd8ee.png)

# Possible solutions

## Follow TypeScript's recommendations 

> Needless to say, I'm personally biased toward this one as it is the easiest one to support :sweat_smile:

The Astro frontmatter behaves like TypeScript, looks like TypeScript and well, is TypeScript. Users used to TypeScript will by default always type their imports with no file extensions, editors will by default auto-complete imports with no extensions inside TypeScript (including inside Astro files)

As such, we should make TypeScript's way of doing things the main and perhaps only way of doing it. If we still want to support `.ts` imports for retro-compatibility, we can at the same time re-enable the error message shown by TypeScript for `.ts` imports inside the editor so users flock naturally towards no extensions

This would also include supporting `.js` imports for `.ts` files, as they're supported in TypeScript, but this can be a side goal as no extensions imports are the currently main recommended ones. This is however [changing in the future](https://github.com/withastro/rfcs/discussions/191#discussioncomment-2690204) because `moduleResolution: 'nodenext'` requires `.js` imports, so we'll need to support those eventually definitely 

### Script tags

Since we support importing with no extensions in dev inside script tags. I think we should _just_ also support it in build and apply the same strategy of pointing users towards no extension or `.js` imports. That way, it's consistent with the frontmatter and JS/TS files

## Alternative: Adapt our tooling

We can alternatively try to bend TypeScript towards supporting `.ts` imports as much as possible. This was a solution we attempted in the past, but it turned out to have many issues due to a quick implementation. There's many ways to support this and one of them could be better then the others:

- Correct the imports inside our TSX output
- Work around the module resolution of TypeScript and redirect `.ts` files to something else
- Try to magically pretend that `.ts` files are something else <--- We tried this

However, I am personally against this solution as it's a lot of work, it introduce additional maintenance, goes against what users are used to and goes against what the technology used recommend. It should still be considered however, as even though imports of non-existing files is something TypeScript recommend, it might not work with other technologies that we're building on

# Links

- [TypeScript module resolution documentation](https://www.typescriptlang.org/docs/handbook/module-resolution.html)

**Comment by jasikpark** [comment]
Something to pay attention to for this is that Typescript is having a renaissance about this themselves with trying to support ESM imports more directly: https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-beta/#esm-nodejs

**Reply by jasikpark** [reply]
https://github.com/microsoft/TypeScript/issues/48027

**Reply by Princesseuh** [reply]
Yes! When using `moduleResolution: 'nodenext'`, the only supported mode is using a `.js` extension, which Astro does not support at all at the moment. (I think this is only relevant at build time when using `tsc` though? The language service APIs still works with no extension even with `nodenext`. Maybe they're not updated yet though)

Definitely something to consider, thanks for bringing attention to it!

**Comment by aFuzzyBear** [comment]
Free the duck from its wellies, let it paddle! 
Aye this makes sense, the arguments for retaining the `.ts` imports feel like its just a maintenance headache that can be avoided.

**Comment by Princesseuh** [comment]
We had an internal talk about this in the team and decided to go toward aligning ourselves with TypeScript and Vite, this is more seen as a bug than a new feature. 

The VS Code extension will warn about `.ts` imports in the future and instead recommend no extensions or `.js` (when the latter is supported)

This should ship as part of 1.0! :partying_face: :tada:
